src/librustc_typeck/check/mod.rs

```diff
@@ -3471,8 +3471,39 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             // We won't diverge unless both branches do (or the condition does).
             self.diverges.set(cond_diverges | then_diverges & else_diverges);
         } else {
+            // If this `if` expr is the parent's function return expr, the cause of the type
```

Comment:
Forgive the slightly annoying but well-meaning intrusion: Can we refactor this into functions?